awx/main/db/profiled_pg/base.py

```diff
@@ -0,0 +1,101 @@
+import os
+import traceback
+import uuid
+
+from django.core.cache import cache
+from django.core.cache.backends.locmem import LocMemCache
+from django.db.backends.postgresql.base import DatabaseWrapper
+
+from awx.main.utils import memoize
+
+__loc__ = LocMemCache(str(uuid.uuid4()), {})
+__all__ = ['DatabaseWrapper']
+
+
+class RecordedQueryLog(object):
+
+    dest = '/var/lib/awx/profile'
+
+    def __init__(self, log, db):
+        self.log = log
+        self.db = db
+        self.threshold = cache.get('awx-profile-sql-threshold')
+
+    def append(self, query):
+        ret = self.log.append(query)
+        self.write(query)
+        return ret
+
+    def write(self, query):
+        if self.threshold is None:
+            return
+        seconds = float(query['time'])
+        if seconds >= self.threshold:
+            sql = query['sql']
+            if sql.startswith('EXPLAIN'):
+                return
+            if not os.path.isdir(self.dest):
+                os.makedirs(self.dest)
+            filename = '%.3fs-%s.sql' % (seconds, uuid.uuid4())
+            filepath = os.path.join(self.dest, filename)
+            bt = ' '.join(traceback.format_stack())
+            explain = ''
+            cursor = self.db.cursor()
+            cursor.execute('EXPLAIN ANALYZE VERBOSE {}'.format(sql))
+            for line in cursor.fetchall():
+                explain += line[0] + '\n'
+            with open(filepath, 'w') as f:
+                f.write('\n\n'.join([sql, explain, bt]))
+
+    def __len__(self):
+        return len(self.log)
+
+    def __iter__(self):
+        return iter(self.log)
+
+    def __getattr__(self, attr):
+        return getattr(self.log, attr)
+
+
+class DatabaseWrapper(DatabaseWrapper):
+    """
+    This is a special subclass of Django's postgres DB backend which - based on
+    the value of a special flag in memcached - captures slow queries and
+    writes profile and Python stack metadata to the disk.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(DatabaseWrapper, self).__init__(*args, **kwargs)
+        # Django's default base wrapper implementation has `queries_log`
+        # which is a `collections.deque` that every query is appended to
+        #
+        # this line wraps the deque with a proxy that can capture each query
+        # and - if it's slow enough - record profiling metadata to the file
+        # system for debugging purposes
+        self.queries_log = RecordedQueryLog(self.queries_log, self)
+
+    @property
+    @memoize(ttl=1, cache=__loc__)
+    def force_debug_cursor(self):
```

Comment:
I don't love storing every django query in memory. We can do better than that. We know we can offload (and do) this information to a file.

Reply:
To achieve this, we can overwrite the django.db.backends.utils.CursorWrapper ourselves. i.e. https://gist.github.com/chrismeyersfsu/f0a5e6109cc134a533c7e9d0b2ba5cb4

Reply:
So Django kind of supports this already.  The queries *are* stored in memory, but the internal `collections.deque` it uses has a default upper bound:

https://github.com/django/django/blob/978ad6d4c043f5004005b673c6e7bb8f415a35dc/django/db/backends/base/base.py#L57

By default, it only keeps the last 9000 queries (we could modify that if we wanted).